File: 16-String-Part-II/python/03_minimum_insertion_steps_to_make_a_string_palindrome.py
class Solution:
    def minInsertions(self, s: str) -> int:
        n = len(s)
        memo = {}

        def dfs(i, j):
            if i >= j:
                return 0
            if (i, j) in memo:
                return memo[(i, j)]

            if s[i] == s[j]:
                res = dfs(i + 1, j - 1)
            else:
                res = min(dfs(i + 1, j), dfs(i, j - 1)) + 1

            memo[(i, j)] = res
            return res

        return dfs(0, n - 1)

File: 16-String-Part-II/cpp/03-minimum-insertion-steps-to-make-a-string-palindrome.cpp
class Solution {
    vector<vector<int>> memo;

public:
    int minInsertions(string s) {
        int n = s.size();
        memo.assign(n, vector<int>(n, -1));
        return dfs(0, n - 1, s);
    }

private:
    int dfs(int i, int j, const string& s) {
        if (i >= j) return 0;
        if (memo[i][j] != -1) return memo[i][j];

        if (s[i] == s[j]) {
            return memo[i][j] = dfs(i + 1, j - 1, s);
        } else {
            return memo[i][j] = min(dfs(i + 1, j, s), dfs(i, j - 1, s)) + 1;
        }
    }
};
